src/Symfony/Component/Messenger/Middleware/DeduplicateMiddleware.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Stamp\DeduplicateStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class DeduplicateMiddleware implements MiddlewareInterface
+{
+    public function __construct(private LockFactory $lockFactory)
+    {
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (!$stamp = $envelope->last(DeduplicateStamp::class)) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        if (!$envelope->last(ReceivedStamp::class)) {
+            $lock = $this->lockFactory->createLockFromKey($stamp->getKey(), $stamp->getTtl(), autoRelease: false);
+
+            if (!$lock->acquire()) {
+                return $envelope;
+            }
+        } elseif ($stamp->onlyDeduplicateInQueue()) {
+            $this->lockFactory->createLockFromKey($stamp->getKey())->release();
+        }
+
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
+            if ($envelope->last(ReceivedStamp::class) && !$stamp->onlyDeduplicateInQueue()) {
+                $this->lockFactory->createLockFromKey($stamp->getKey())->release();
+            }
+        }
```

Comment:
I'm not sure what you mean by dispatching.

To me there is three step
- I put the message in queue
- I receive the message from the queue to treat the message
- I finished the treatment

Only in queue mean I don't want to put two times the same message in queue because they will have the same effect. But since the treatment of the message take time and has multiple steps, even if a first message is in treatment, the following might have a different result so I have to add it to the queue.

The other one, means as soon as I have a message in queue or in treatment, I don't want to add another one.